internal/mempool/types.go

```diff
@@ -23,45 +23,57 @@ const (
-type Mempool interface {
+// Status is the status of the Pool.
+type Status string
+
+const (
+	// StatusTXsAvailable indicates the pool has txs available.
+	StatusTXsAvailable = "txs_available"
+)
+
+type MempoolABCI interface {
 	// CheckTx executes a new transaction against the application to determine
 	// its validity and whether it should be added to the mempool.
 	CheckTx(ctx context.Context, tx types.Tx, callback func(*abci.ResponseCheckTx), txInfo TxInfo) error
 
-	// RemoveTxByKey removes a transaction, identified by its key,
-	// from the mempool.
-	RemoveTxByKey(txKey types.TxKey) error
+	// EnableTxsAvailable initializes the TxsAvailable channel, ensuring it will
+	// trigger once every height when transactions are available.
+	EnableTxsAvailable()
 
-	// ReapMaxBytesMaxGas reaps transactions from the mempool up to maxBytes
-	// bytes total with the condition that the total gasWanted must be less than
-	// maxGas.
-	//
-	// If both maxes are negative, there is no cap on the size of all returned
-	// transactions (~ all available transactions).
-	ReapMaxBytesMaxGas(maxBytes, maxGas int64) types.Txs
+	// Flush will clear all caches in the MempoolABCI and clear any volatile
+	// memory within the pool.
+	Flush(ctx context.Context) error
 
-	// ReapMaxTxs reaps up to max transactions from the mempool. If max is
-	// negative, there is no cap on the size of all returned transactions
-	// (~ all available transactions).
-	ReapMaxTxs(max int) types.Txs
+	// PoolMeta returns the metadata for the underlying pool implementation.
+	PoolMeta() PoolMeta
 
-	// Lock locks the mempool. The consensus must be able to hold lock to safely
-	// update.
-	Lock()
+	// PrepBlockFinality prepares the mempool for finalizing a block. This may require
+	// locking or other concerns that must be handled before a block is safe to commit.
+	PrepBlockFinality(ctx context.Context) (finishFn func(), err error)
 
-	// Unlock unlocks the mempool.
-	Unlock()
+	// Reap returns Txs from the given pool. It is up to the pool implementation to define
+	// how they handle the possible predicates from option combinations.
+	Reap(ctx context.Context, opts ...ReapOptFn) (types.Txs, error)
+
+	// Remove removes Txs from the given pool. It is up to the pool implementation to
+	// define how they handle the possible predicates from option combinations.
+	Remove(ctx context.Context, opts ...RemOptFn) error
+
+	// TxsAvailable returns a channel which fires once for every height, and only
+	// when transactions are available in the mempool.
+	//
+	// NOTE:
+	// 1. The returned channel may be nil if EnableTxsAvailable was not called.
+	TxsAvailable() <-chan struct{}
 
 	// Update informs the mempool that the given txs were committed and can be
 	// discarded.
 	//
 	// NOTE:
 	// 1. This should be called *after* block is committed by consensus.
-	// 2. Lock/Unlock must be managed by the caller.
+	// 2. PrepBlockFinality must be called before calling Update, effectively
```

Comment:
Numbered lists in godoc should be indented to render correctly: https://go.dev/doc/comment#lists